src/basicfuns.jl

```diff
@@ -232,20 +232,33 @@ function logsumexp(X)
     isempty(X) && return log(sum(X))
     reduce(logaddexp, X)
 end
+
+"""
+    logsumexp(X::AbstractArray, [W]; dims=:)
+
+Compute `log(sum(exp, X))`, evaluated avoiding intermediate overflow/undeflow along the specified dimension. `W` is an optional array of weights to apply to the sum, effectively `log(sum((x,w) -> w * exp(x), zip(X,W)))`.
+"""
 function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     # Do not use log(zero(T)) directly to avoid issues with ForwardDiff (#82)
     u = reduce(max, X, dims=dims, init=oftype(log(zero(T)), -Inf))
     u isa AbstractArray || isfinite(u) || return float(u)
-    let u=u # avoid https://github.com/JuliaLang/julia/issues/15276
-        # TODO: remove the branch when JuliaLang/julia#31020 is merged.
-        if u isa AbstractArray
-            u .+ log.(sum(exp.(X .- u); dims=dims))
-        else
-            u + log(sum(x -> exp(x-u), X))
-        end
+    if u isa AbstractArray
+        u .+ log.(sum(exp.(X .- u); dims=dims))
+    else
+        u + log(sum(x -> exp(x-u), X))
     end
 end
 
+function logsumexp(X::AbstractArray{T}, W; dims=:) where {T<:Real}
```

Comment:
in the docstring you say `W` is an array, but there is no type restriction in the signature. I think it would be cleaner to have it especially that in your code `W` and `X` do not have to have matching dimensions, and when do not then `zip(X, W)` will not iterate every element of `X` in corner cases.

Reply:
added words and examples

Reply:
@nalimilan I didn't want to over-specialize because then I couldn't use a call like `logsumexp([1, 2], (0.5, 0.3))`.

As for the making the sizes the same, that would disallow broadcasting like
`logsumexp(randn(3, 4), rand(1, 4), dims=1)`. By not adding a size check it's more flexible.

Perhaps your concerns could be alleviated by thinking about combining some of David's work in #97 here. The only reason I latch onto this array method when my use case will be tuples is that the non-array method currently is pretty slow for my use.

Reply:
> @nalimilan I didn't want to over-specialize because then I couldn't use a call like `logsumexp([1, 2], (0.5, 0.3))`. This is useful for me because my weights are static in the calculations I need so I prefer to use a Tuple.

Could you use StaticArrays instead of tuples? I tend to prefer restrictive APIs unless we have a strong reason to be more flexible, as we can easily end up allowing things that weren't intended, and which can even return incorrect results.


> As for the making the sizes the same, that would disallow broadcasting like
> `logsumexp(randn(3, 4), rand(1, 4), dims=1)`. By not adding a size check it's more flexible.

As I said the problem is that the definition would be inconsistent with StatsBase, which we should really avoid. With StatsBase conventions, `logsumexp(randn(3, 4), rand(1, 4), dims=1)` would be written as just `logsumexp(randn(3, 4), rand(4), dims=1)`, but it's a bit more complicated to implement. So again unless there's a strong reason

> Perhaps your concerns could be alleviated by thinking about combining some of David's work in #97 here. The only reason I latch onto this array method when my use case will be tuples is that the non-array method currently is pretty slow for my use.

That wouldn't really change anything AFAICT. With non-`AbstractArray` weights you can't check the size of the input to make sure it makes sense and that there's a one-to-one correspondence between values and weights.

Reply:
Okay, well I don't really agree that this is better, but I've fully restricted the function to only work with abstract arrays which have exactly the same size. I've updated the tests and docstring, too.